service/handlers.go

```diff
@@ -107,82 +91,78 @@ func VersionHandler(w http.ResponseWriter, r *http.Request) {
 
 // SignHandler is the API method to sign assertions from the device
 func SignHandler(w http.ResponseWriter, r *http.Request) {
-	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
-
 	if r.Body == nil {
 		w.WriteHeader(http.StatusBadRequest)
-		formatSignResponse(false, "error-nil-data", "", "Uninitialized POST data", "", w)
+		formatSignResponse(false, "error-nil-data", "", "Uninitialized POST data", nil, w)
 		return
 	}
 
-	assertions := new(Assertions)
-	err := json.NewDecoder(r.Body).Decode(&assertions)
-
-	defer r.Body.Close()
-
-	switch {
-	// Check we have some data
-	case err == io.EOF:
+	// Read the full request body
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
 		w.WriteHeader(http.StatusBadRequest)
-		formatSignResponse(false, "error-sign-empty", "", "No data supplied for signing", "", w)
+		formatSignResponse(false, "error-sign-read", "", err.Error(), nil, w)
 		return
-		// Check for parsing errors
-	case err != nil:
+	}
+	if len(data) == 0 {
 		w.WriteHeader(http.StatusBadRequest)
-		errorMessage := fmt.Sprintf("%v", err)
-		formatSignResponse(false, "error-decode-json", "", errorMessage, "", w)
+		formatSignResponse(false, "error-sign-empty", "", "No data supplied for signing", nil, w)
 		return
 	}
 
-	// Validate the model by checking that it exists on the database
-	model, err := Environ.DB.FindModel(assertions.Brand, assertions.Model, assertions.Revision)
+	defer r.Body.Close()
+
+	// Use the ubuntu-core assertions module to decode the body and validate
+	assertion, err := asserts.Decode(data)
 	if err != nil {
-		w.WriteHeader(http.StatusNotFound)
-		formatSignResponse(false, "error-model-not-found", "", "Cannot find model with the matching brand, model and revision", "", w)
+		w.WriteHeader(http.StatusBadRequest)
+		formatSignResponse(false, "error-decode-json", "", err.Error(), nil, w)
 		return
 	}
 
-	// Format the assertions string
-	dataToSign, err := formatAssertion(assertions)
-	if err != nil {
+	// Check that we have a device-serial assertion (the details will have been validated by Decode call)
+	if assertion.Header("type") != "device-serial" {
 		w.WriteHeader(http.StatusBadRequest)
-		errorMessage := fmt.Sprintf("%v", err)
-		formatSignResponse(false, "error-format-assertions", "", errorMessage, "", w)
+		formatSignResponse(false, "error-decode-assertion", "error-invalid-type", "The assertion type must be 'device-serial'", nil, w)
 		return
 	}
 
-	// Read the private key into a string using the model's signing key
-	privateKey, err := getPrivateKey(model.SigningKey)
+	revision, err := strconv.Atoi(assertion.Header("revision"))
```

Comment:
This check was left as Decode will validate the revision only if it is supplied in the assertion. However, if the revision tag is missing from the assertion, Decode misses that fact and defaults the revision to 0. This will get caught by the FindModel call, which won't find the model.

The TestSignHandlerBadAssertionNoRevision test was there for this scenario, but will need a tweak as my mock FindModel always found the model. I'll change that to do a more realistic check of the supplied parameters.